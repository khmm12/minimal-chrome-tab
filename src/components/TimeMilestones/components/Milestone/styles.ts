import { css } from 'styled-system/css'
import * as p from 'styled-system/patterns'

export const container = p.vstack.raw({
  alignItems: 'center',
  gap: 0,
})

export const value = css.raw({
  fontSize: { base: '6.4rem', smDown: '5.4rem' },
})

export const description = css.raw({
  margin: '1rem 0 0',
  fontSize: { base: '2.4rem', smDown: '1.6rem' },
  textTransform: 'lowercase',
})
