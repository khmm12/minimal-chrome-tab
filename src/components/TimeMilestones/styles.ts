import { css } from '@linaria/core'
import { beforeSmall } from '@/theme/media'

export const container = css`
  align-self: flex-start;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  justify-content: center;

  @media ${beforeSmall} {
    align-items: center;
  }
`

export const title = css`
  font-size: 3.6rem;
  margin: 0 0 2.4rem;

  @media ${beforeSmall} {
    font-size: 3.2rem;
    text-align: center;
    margin: 0 0 1.6rem;
  }
`

export const items = css`
  display: grid;
  width: 100%;
  grid-gap: 3.5rem;
  grid-template-columns: repeat(auto-fit, minmax(12rem, auto));
  justify-content: flex-end;

  @media ${beforeSmall} {
    justify-content: center;
    grid-gap: 1.6rem;
  }
`
