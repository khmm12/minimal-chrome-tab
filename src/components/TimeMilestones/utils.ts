import {
  differenceInSeconds,
  getDaysInMonth,
  getDaysInYear,
  startOfDay,
  startOfISOWeek,
  startOfMonth,
  startOfYear,
  getYear,
  setYear,
} from 'date-fns'
import { BirthDate } from '@/config'
import round from '@/utils/round'

const PRECISION = 2

type GetSecondsIn = (dateTime: Date) => number
type GetStartOf = (dateTime: Date) => Date
type GetMilestone = (dateTime: Date) => number

const getMinutesInHour = constant(60)
const getHoursInDay = constant(24)
const getDaysInWeek = constant(7)

const getSecondsInMinute = constant(60)
const getSecondsInHour = mul(getMinutesInHour, getSecondsInMinute)
const getSecondsInDay = mul(getHoursInDay, getSecondsInHour)
const getSecondsInWeek = mul(getDaysInWeek, getSecondsInDay)
const getSecondsInMonth = mul(getDaysInMonth, getSecondsInDay)
const getSecondsInYear = mul(getDaysInYear, getSecondsInDay)

export const getDayMilestone = makeMilestoneCalculator(getSecondsInDay, startOfDay)
export const getWeekMilestone = makeMilestoneCalculator(getSecondsInWeek, startOfISOWeek)
export const getMonthMilestone = makeMilestoneCalculator(getSecondsInMonth, startOfMonth)
export const getYearMilestone = makeMilestoneCalculator(getSecondsInYear, startOfYear)
export const getBirthDayMilestone = makeMilestoneCalculator(getSecondsInBirthDayYear, getPreviousBirthDate)

function getSecondsInBirthDayYear(dateTime: Date): number {
  const getSecondsInYearDate = makeGetSecondsInYearDate(new Date(BirthDate))
  return getSecondsInYearDate(dateTime)
}

function getPreviousBirthDate(dateTime: Date): Date {
  const getPreviousYearDate = makeGetPreviousYearDate(new Date(BirthDate))
  return getPreviousYearDate(dateTime)
}

function makeGetSecondsInYearDate(yearDate: Date): GetSecondsIn {
  const getPreviousYearDate = makeGetPreviousYearDate(yearDate)
  return function getSecondsInYearDate(dateTime: Date) {
    const previousYearDate = getPreviousYearDate(dateTime)
    const nextYearDate = setYear(previousYearDate, getYear(previousYearDate) + 1)
    return differenceInSeconds(nextYearDate, previousYearDate)
  }
}

function makeGetPreviousYearDate(yearDate: Date) {
  return function getPreviousYearDate(dateTime: Date): Date {
    const startOfGivenDay = startOfDay(dateTime)
    const startOfYearDateDay = startOfDay(yearDate)
    const givenYear = getYear(startOfGivenDay)

    const previousGivenYearDate = setYear(startOfYearDateDay, givenYear - 1)
    const givenYearDate = setYear(startOfYearDateDay, givenYear)
    return givenYearDate < startOfGivenDay ? givenYearDate : previousGivenYearDate
  }
}

function makeMilestoneCalculator(getSecondIn: GetSecondsIn, getStartOf: GetStartOf): GetMilestone {
  return function milestoneCalculator(dateTime) {
    const passedSeconds = differenceInSeconds(dateTime, getStartOf(dateTime))
    const value = passedSeconds / getSecondIn(dateTime)
    return round(value, PRECISION)
  }
}

function constant<T>(value: T): () => T {
  return () => value
}

function mul(...fns: GetSecondsIn[]): GetSecondsIn {
  return (dateTime) => fns.reduce((result, fn) => result * fn(dateTime), 1)
}
