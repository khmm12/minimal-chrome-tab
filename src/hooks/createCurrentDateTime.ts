import { Accessor, createSignal } from 'solid-js'
import createInterval, { Every } from '@/hooks/createInterval'
import useTabActive from '@/hooks/useTabActive'

interface CreateCurrentDateTimeConfig {
  updateEvery?: Every
}

const getDate = (): Date => new Date()

export default function createCurrentDateTime(config?: CreateCurrentDateTimeConfig): Accessor<Date> {
  const [dateTime, setDateTime] = createSignal(getDate())
  const update = (): Date => setDateTime(getDate())

  const isTabActive = useTabActive()

  createInterval(update, {
    get every() {
      return config?.updateEvery ?? 'second'
    },
    get enabled() {
      return isTabActive()
    },
  })

  return dateTime
}
