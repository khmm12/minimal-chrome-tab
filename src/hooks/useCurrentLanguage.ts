import type { Accessor } from 'solid-js'
import getCurrentLocale from '@/utils/get-current-locale'

export default function useCurrentLanguage(): Accessor<string> {
  const language = getCurrentLocale()
  return () => language
}
