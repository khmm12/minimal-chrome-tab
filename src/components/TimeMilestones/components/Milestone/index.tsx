import type { JSX } from 'solid-js'
import { css } from 'styled-system/css'
import createIntlFormatter from '@/hooks/createIntlFormatter'
import { round } from '@/utils/rounds'
import Bar from './components/Bar'
import * as s from './styles'

interface MilestoneProps {
  value: number
  description: string
}

const PercentFormatOptions = {
  style: 'percent',
  minimumFractionDigits: 0,
  maximumFractionDigits: 0,
} satisfies Intl.NumberFormatOptions

export default function Milestone(props: MilestoneProps): JSX.Element {
  const format = createIntlFormatter()

  const formatPercent = (value: number): string => format.number(value, PercentFormatOptions)
  const formatValue = (value: number): string => formatPercent(round(value, 2)).replaceAll(/\s/g, '')

  return (
    <div role="group" aria-label={props.description} class={css(s.container)}>
      <span class={css(s.value)}>{formatValue(props.value)}</span>
      <Bar progress={props.value} />
      <span class={css(s.description)}>
        <span aria-hidden="true">...</span>
        {props.description}
      </span>
    </div>
  )
}
