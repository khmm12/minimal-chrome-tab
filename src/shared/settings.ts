import * as v from 'valibot'
import { isISODate, type ISODate } from '@/utils/brands'
import Storage, { buildStorageAdapter, type ISerializer } from '@/utils/storage'

export interface Settings {
  birthDate?: ISODate
  themeColorMode: ThemeColorMode
  milestoneProgressStyle: MilestoneProgressStyle
}

export enum ThemeColorMode {
  Auto = 'auto',
  Light = 'light',
  Dark = 'dark',
}

export enum MilestoneProgressStyle {
  BarsCompact = 'bars-compact',
  BarsDetailed = 'bars-detailed',
  HorizontalBar = 'bar',
}

const Key = 'settings'

const isoDateSchema = /* @__PURE__ */ v.custom<ISODate>(isISODate, 'is not a valid ISODate')

const defaults = {
  milestoneProgressStyle: MilestoneProgressStyle.BarsCompact,
  themeColorMode: ThemeColorMode.Auto,
} satisfies Settings

const SettingsSchema = /* @__PURE__ */ v.fallback(
  v.object({
    birthDate: v.fallback(v.optional(isoDateSchema), undefined),
    milestoneProgressStyle: v.fallback(
      v.optional(v.enum(MilestoneProgressStyle), defaults.milestoneProgressStyle),
      defaults.milestoneProgressStyle,
    ),
    themeColorMode: v.fallback(v.optional(v.enum(ThemeColorMode), defaults.themeColorMode), defaults.themeColorMode),
  }),
  defaults,
)

const Serializer: ISerializer<Settings> = {
  deserialize(value) {
    try {
      return v.parse(SettingsSchema, value)
    } catch {
      return v.getFallback(SettingsSchema)
    }
  },
  serialize(value) {
    return value
  },
}

const StorageAdapter = /* @__PURE__ */ await buildStorageAdapter(Key)

// Singletone
export const SettingsStorage = /* @__PURE__ */ await Storage.create(StorageAdapter, Serializer)
