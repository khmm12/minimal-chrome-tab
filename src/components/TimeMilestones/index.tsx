import { createMemo, type JSX, Show } from 'solid-js'
import { css, cx } from 'styled-system/css'
import createCurrentDateTime, { EveryClockMinute } from '@/hooks/createCurrentDateTime'
import createUniqueIds from '@/hooks/createUniqueIds'
import { MilestoneProgressStyle } from '@/shared/settings'
import asGetters from '@/utils/as-getters'
import type { ISODate } from '@/utils/brands'
import Milestone, { MilestoneVariant } from './components/Milestone'
import createTimeMilestones from './hooks/createTimeMilestones'
import * as s from './styles'

export interface TimeMilestonesProps {
  birthDate?: ISODate | undefined
  progressStyle: MilestoneProgressStyle
}

export default function TimeMilestones(props: TimeMilestonesProps): JSX.Element {
  const currentDateTime = createCurrentDateTime({ update: EveryClockMinute })

  const birthDate = createMemo(() => createDate(props.birthDate), null, { equals: isDateEqual })
  const variant = (): MilestoneVariant => mapSettingsStyleToVariant(props.progressStyle)

  const milestones = createTimeMilestones(asGetters({ currentDateTime, birthDate }))

  const ids = createUniqueIds(['heading'])

  return (
    <div role="group" aria-describedby={ids.heading} aria-label="Time milestones" class={css(s.container)}>
      <h1 id={ids.heading} class={css(s.title)}>
        We're now through...
      </h1>
      <div class={cx(css(s.items))}>
        <Milestone variant={variant()} value={milestones.day} description="of day" />
        <Milestone variant={variant()} value={milestones.week} description="of week" />
        <Milestone variant={variant()} value={milestones.month} description="of month" />
        <Milestone variant={variant()} value={milestones.year} description="of year" />
        <Show when={milestones.birthday}>
          {(v) => <Milestone variant={variant()} value={v()} description="of b'day" />}
        </Show>
      </div>
    </div>
  )
}

function createDate(d: ISODate | undefined): Date | null {
  return d != null ? new Date(d) : null
}

function isDateEqual<T extends Date | undefined | null>(a: T, b: T): boolean {
  return a?.valueOf() === b?.valueOf()
}

function mapSettingsStyleToVariant(style: MilestoneProgressStyle): MilestoneVariant {
  switch (style) {
    case MilestoneProgressStyle.BarsCompact:
      return MilestoneVariant.BarsCompact
    case MilestoneProgressStyle.BarsDetailed:
      return MilestoneVariant.BarsDetailed
    case MilestoneProgressStyle.HorizontalBar:
      return MilestoneVariant.HorizontalBar
  }
}
