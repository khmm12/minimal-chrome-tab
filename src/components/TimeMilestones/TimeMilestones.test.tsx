import { render, screen } from '@solidjs/testing-library'
import createCurrentDateTime from '@/hooks/createCurrentDateTime'
import useCurrentLanguage from '@/hooks/useCurrentLanguage'
import { MilestoneProgressStyle } from '@/shared/settings'
import toISODate from '@/utils/to-iso-date'
import Milestone from './components/Milestone'
import TimeMilestones, { type TimeMilestonesProps } from '.'

vi.mock('@/hooks/useCurrentLanguage')
vi.mock('@/hooks/createCurrentDateTime')
vi.mock('./components/Milestone')

beforeEach(() => {
  vi.mocked(useCurrentLanguage).mockReturnValue(() => 'en-GB')
  vi.mocked(createCurrentDateTime).mockReturnValue(() => new Date('2022-03-05T16:05:30'))
  vi.mocked(Milestone).mockImplementation((props) => (
    <div aria-label={props.description}>
      {props.description} {props.value}
    </div>
  ))
})

afterEach(() => {
  vi.resetAllMocks()
})

describe('TimeMilestones', () => {
  it('renders correctly', () => {
    render(() => <TimeMilestones {...defaults()} />)

    const r = screen.getByLabelText('Time milestones')

    expect(r).toBeInTheDocument()
    expect(r).toMatchSnapshot()
  })

  it('has heading', () => {
    render(() => <TimeMilestones progressStyle={MilestoneProgressStyle.BarsCompact} />)

    const heading = screen.getByText("We're now through...")
    expect(heading).toBeInTheDocument()
  })

  it('has main milestones', () => {
    render(() => <TimeMilestones progressStyle={MilestoneProgressStyle.BarsCompact} />)

    const ofDayMileStone = screen.getByText(/of day/)
    expect(ofDayMileStone).toBeInTheDocument()

    const ofWeekMileStone = screen.getByText(/of week/)
    expect(ofWeekMileStone).toBeInTheDocument()

    const ofMonthMileStone = screen.getByText(/of month/)
    expect(ofMonthMileStone).toBeInTheDocument()
  })

  it('has optional birthday milestone if specified', () => {
    render(() => <TimeMilestones {...defaults({ birthDate: toISODate('1970-01-01') })} />)

    const ofBirthdayMileStone = screen.getByText(/of b'day/)
    expect(ofBirthdayMileStone).toBeInTheDocument()
  })

  it('has no birthday milestone if not specified', () => {
    render(() => <TimeMilestones {...defaults({ birthDate: undefined })} />)
    const ofBirthdayMileStone = screen.queryByText(/of b'day/)
    expect(ofBirthdayMileStone).not.toBeInTheDocument()
  })
})

function defaults(opts?: Partial<TimeMilestonesProps>): TimeMilestonesProps {
  return {
    progressStyle: MilestoneProgressStyle.BarsCompact,
    ...opts,
  }
}
