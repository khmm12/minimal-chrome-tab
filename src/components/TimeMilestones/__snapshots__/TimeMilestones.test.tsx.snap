// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TimeMilestones > renders correctly 1`] = `
<div
  aria-describedby="cl-0"
  aria-label="Time milestones"
  class="d_flex ai_flex-end smDown:ai_center jc_center gap_0 flex-d_column as_flex-start"
  role="group"
>
  <h1
    class="fs_3.6rem smDown:fs_3.2rem fw_bold ta_center m_0_0_2.4rem smDown:m_0_0_1.6rem"
    id="cl-0"
  >
    We're now through...
  </h1>
  <div
    class="d_grid grid-tc_repeat(auto-fit,_minmax(12rem,_auto)) gap_3.5rem smDown:gap_1.6rem w_full jc_flex-end smDown:jc_center"
  >
    <div
      aria-label="of day"
    >
      of day
       
      0.6704
    </div>
    <div
      aria-label="of week"
    >
      of week
       
      0.81
    </div>
    <div
      aria-label="of month"
    >
      of month
       
      0.1506
    </div>
    <div
      aria-label="of year"
    >
      of year
       
      0.1744
    </div>
    
  </div>
</div>
`;
